### ISSUE TITLE
Polymer Behaviors Not Compatible with Type Based Optimizations

### ISSUE THREAD

**Issue by ChadKillingsworth** [google/closure-compiler#2119]
I've yet to figure out how to create a simple repo, but the type based optimizations can break Polymer behaviors. The behavior literal keys are renamed in a conflicting manner:

```js
/** @polymerBehavior */
var fooBehavior = {
  foo: function() { alert('foo'); },
};

/** @polymerBehavior */
var barBehavior = {
  bar: function() { alert('bar'); },
};

(function() {
  Polymer({
    is: 'foo-bar',
    behaviors: [fooBehavior, barBehavior]
  });
})();
```

Becomes something like:

```js
/** @polymerBehavior */
var a = {
  b: function() { alert('foo'); },
};

/** @polymerBehavior */
var c = {
  b: function() { alert('bar'); },
};

(function() {
  Polymer({
    is: 'foo-bar',
    behaviors: [a, c]
  });
})();
```

I believe the way to fix this is to change the PolymerPass processing of behaviors such that a synthetic interface is created for each behavior. The behavior object literal will then be type cast to that interface and any Polymer elements using the behavior will be marked as implementing it.

I'm guessing this hasn't come up at Google because the Polymer-Renamer is already incompatible with type based optimizations. However https://github.com/banno/polymer-rename does support them.

**Comment by dimvar** [comment]
AFAIU, you're saying that the root of the problem is that the two object literals appear unrelated to the type-based optimizations, but are combined in `behaviors: [a, c]`, so their properties should have distinct names.

However, the way property disambiguation works, properties on object literals are the same as defining a property on `Object`, so they are related to all object types, and are never disambiguated. (Example)[https://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520o1%2520%253D%2520%257B%2520mypropone%253A%2520123%2520%257D%253B%250A%250Avar%2520o2%2520%253D%2520%257B%2520myproptwo%253A%2520234%2520%257D%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26DISAMBIGUATE_PROPERTIES%3D1%26GENERATE_PSEUDO_NAMES%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26PROPERTY_RENAMING%3D1%26TRANSPILE%3D1]

So, either the polymer pass rewrites them in some weird way (I'm not very familiar with the polymer pass), or the bug is in AmbiguateProperties. You can try editing the source to turn on/off individual passes and enable pseudo renaming, to see if you can narrow it down.

**Comment by ChadKillingsworth** [comment]
> or the bug is in AmbiguateProperties

This appears to be the case. I created a custom build with ambiguateProperties turned off, but the other type-based optimizations enabled and the issue didn't occur.

Debugging further to find out why.

**Comment by MatrixFrog** [comment]
cc @jklein24 @rictic

**Comment by ChadKillingsworth** [comment]
More info: while we avoid disambiguating properties on object literals, we currently do ambiguate them - thus the issue. This has worked well for a long time and I'd hate to change it outright.

I think I'd like to change the polymer pass such that polymer behaviors are marked as records and polymer elements marked as implementing those records. I believe that will fix the issue.

**Comment by ChadKillingsworth** [comment]
This was closed by #2143
